impl/src/main/java/org/glassfish/soteria/authorization/spi/impl/SubjectParser.java

```diff
@@ -389,181 +383,200 @@ private boolean tryWebLogic(String contextID, Collection<String> allDeclaredRole
                 return groups;
             }
         }
- 
+
         return groups;
     }
-     
+
     public List<String> principalToGroups(Principal principal) {
         List<String> groups = new ArrayList<>();
         principalToGroups(principal, groups);
         return groups;
     }
-    
+
     private Principal doGetCallerPrincipalFromPrincipals(Iterable<Principal> principals) {
-        
         for (Principal principal : principals) {
-            switch (principal.getClass().getName()) {
-                case "org.glassfish.security.common.PrincipalImpl": // GlassFish/Payara
-                // JBoss EAP/WildFly convention 1 - single top level principal of the below type
-                case "org.jboss.security.SimplePrincipal":
-                case "com.ibm.ws.security.authentication.principals.WSPrincipal": // Liberty
-                    return principal;
-                // JBoss EAP/WildFly convention 2 - the one and only principal in group called CallerPrincipal
-                case "org.jboss.security.SimpleGroup": 
-                    if (principal.getName().equals("CallerPrincipal") && principal instanceof Group) {
-                        
-                        Enumeration<? extends Principal> groupMembers = ((Group) principal).members();
-                        
-                        if (groupMembers.hasMoreElements()) {
-                            return groupMembers.nextElement();
-                        }
-                    }
-                    break;
-                case "org.apache.tomee.catalina.TomcatSecurityService$TomcatUser": // TomEE
-                    try {
-                        return 
-                            (Principal) Class.forName("org.apache.catalina.realm.GenericPrincipal")
-                                             .getMethod("getUserPrincipal")
-                                             .invoke(
-                                                 Class.forName("org.apache.tomee.catalina.TomcatSecurityService$TomcatUser")
-                                                      .getMethod("getTomcatPrincipal")
-                                                      .invoke(principal))
-                            
-                            ;
-                        } catch (Exception e) {
-                            
-                        }
-                    break;
-            }
-
+            // If the principal is a standard one, return it
             if (CallerPrincipal.class.isAssignableFrom(principal.getClass())) {
                 return principal;
             }
-            
+
+            // Do some checks to determine it from vendor specific data
+            Principal vendorCallerPrincipal = getVendorCallerPrincipal(principal);
+            if (vendorCallerPrincipal != null) {
+                return vendorCallerPrincipal;
+            }
+
+            // Check for Servlet
```

Comment:
I think we should really delegate to `HttpServletRequest` first, then to `EJBContext`, and only *after* we've tried those two mechanisms should we go looking for another principal. These methods will, by definition, be the "best evidence" for the caller principal. They'll return the correct principal on any platform, and the SecurityContext.getCallerPrincipal() method is defined in terms of the container-specific methods. (We seem to have lost the spec language that says this, but my recollection is the method was originally defined in terms similar to isCallerInRole(), i.e., in terms of the behavior of the existing container APIs. All in all, it seems safer to check there first, and only then start making guesses based on principal types.